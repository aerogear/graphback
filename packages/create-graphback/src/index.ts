#!/usr/bin/env node

import * as yargs from 'yargs';

import { init } from './init/components';

type Params = {
  name?: string;
  templateName?: string;
  templateUrl?: string;
  filter?: string;
};

const command = '$0 <name>';

export const desc = 'Create Graphback project from available templates';

// tslint:disable-next-line: typedef
export const builder = (args: yargs.Argv) => {
  args.positional('name', {
    describe: 'Project name',
    type: 'string'
  });
  args.option('templateName', {
    describe: 'Name of the predefined template',
    type: 'string'
  });
  args.option('templateUrl', {
    describe:
      'GitHub URL of the template. For example (https://github.com/wtrocki/graphql-serve-example#master)',
    type: 'string'
  });
  args.option('filter', {
    describe:
      'A filter option used to limit the number of displayed templates that only matches it. For example "postgres" will only display templates that have a Postgres backed',
    type: 'string'
  });
};

export async function handler({
  name,
  templateName,
  templateUrl,
  filter
}: Params) {
  await init(name, templateName, templateUrl, filter);
}

if (require.main === module) {
  // eslint-disable-next-line no-unused-expressions
  yargs
    .command(command, desc, builder, handler)
    .example(
      '$0 my-awesome-project',
      'creates the "my-awesome-project" in the current working directory'
    )
    .demandCommand(1)
    .strict()
    .recommendCommands()
    .help()
    .alias('h', 'help')
    .version()
    .alias('v', 'version').argv;
}
