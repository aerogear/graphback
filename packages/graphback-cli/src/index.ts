#!/usr/bin/env node

import * as yargs from 'yargs';

if (require.main === module) {
  // eslint-disable-next-line no-unused-expressions
  yargs
    .commandDir('commands')
    .demandCommand(1)
    .strict()
    .recommendCommands()
    .help()
    .alias('h', 'help')
    .version()
    .alias('v', 'version')
    .argv;
}
