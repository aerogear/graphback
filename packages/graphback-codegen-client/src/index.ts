export * from './ClientDocuments'
export * from './ClientCRUDPlugin'