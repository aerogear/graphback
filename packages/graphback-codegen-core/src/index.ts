 
// Top level API
export * from './api'
export * from './graphql'
