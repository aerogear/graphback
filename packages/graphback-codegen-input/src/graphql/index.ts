
export * from "./GraphQLContextCreator"
export * from "./graphqlUtils"