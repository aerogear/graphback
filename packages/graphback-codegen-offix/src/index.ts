export * from './OffixPlugin'
export * from './OffixRuntimeData'


//Required for plugins
export { OffixPlugin as Plugin } from './OffixPlugin'