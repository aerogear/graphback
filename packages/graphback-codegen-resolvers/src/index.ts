export * from "./api/OutputResolver"
export * from "./api/generateResolvers"
export * from './api/ResolverGeneratorOptions'