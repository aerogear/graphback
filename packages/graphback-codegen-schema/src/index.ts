export * from './writer/schemaFormatters'
export * from './SchemaCRUDPlugin'