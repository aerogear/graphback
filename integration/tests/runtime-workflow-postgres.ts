/* eslint-disable no-null/no-null */
/* eslint-disable no-shadow */
/* eslint-disable @typescript-eslint/no-use-before-define */
/* eslint-disable import/no-extraneous-dependencies */
import { mkdirSync, readFileSync, rmdirSync } from 'fs';
import * as path from 'path';
import { ApolloServer } from "apollo-server";
import { createTestClient, ApolloServerTestClient } from 'apollo-server-testing';
import { loadConfig } from 'graphql-config';
import { loadDocuments } from '@graphql-toolkit/core';
import { GraphQLFileLoader } from '@graphql-toolkit/graphql-file-loader';
import * as Knex from 'knex';
import { buildGraphbackAPI, GraphbackAPI } from "graphback/src";
import { DocumentNode } from 'graphql';
import { migrateDB } from '../../packages/graphql-migrations/src';
import { createKnexDbProvider } from "../../packages/graphback-runtime-knex";
import { SchemaCRUDPlugin } from '../../packages/graphback-codegen-schema';
import { ClientCRUDPlugin } from '../../packages/graphback-codegen-client';

/** global config */
let db: Knex;
let server: ApolloServer;
let client: ApolloServerTestClient;
let graphbackApi: GraphbackAPI;

let documents: DocumentNode;

const modelText = readFileSync('./mock.graphql').toString();

beforeAll(async () => {
  try {
    mkdirSync("./output-postgres");
    mkdirSync("./output-postgres/client")
    const dbMigrationsConfig = {
      client: "pg",
      connection: {
        user: "postgresql",
        password: "postgres",
        database: "users",
        host: "localhost",
        port: process.env.PORT || 5432
      }
    }
    db = Knex(dbMigrationsConfig);

    graphbackApi = buildGraphbackAPI(modelText, {
      dataProviderCreator: createKnexDbProvider(db),
      plugins: [
        new SchemaCRUDPlugin({outputPath: "./output-postgres/schema/schema.graphql"}),
        new ClientCRUDPlugin({format: 'graphql', outputFile: './output-postgres/client/graphback.graphql'})
      ]
    });

    const { newDB } = await migrateDB(dbMigrationsConfig, graphbackApi.schema);

    await seedDatabase(db);

    expect(newDB).toMatchSnapshot();

    const source = await loadDocuments(path.resolve(`./output-postgres/client/**/*.graphql`), {
      loaders: [
        new GraphQLFileLoader()
      ]
    });
    documents = source[0].document;
  } catch (e) {
    console.log(e);
    throw e;
  }
})

beforeEach(() => {
  const { typeDefs, resolvers, contextCreator } = graphbackApi;
  server = new ApolloServer({
    typeDefs,
    resolvers,
    context: contextCreator
  });

  client = createTestClient(server);
})

afterEach(() => server.stop())

afterAll(async () => {
  rmdirSync(path.resolve('./output-postgres'), { recursive: true });
  await db.schema.dropTableIfExists('comment').dropTableIfExists('commentmetadata').dropTableIfExists('note');
  return db.destroy();
});

async function seedDatabase(db: Knex) {
  await db('note').insert([
    {
      title: 'Note A',
      description: 'Note A Description'
    },
    {
      title: 'Note B',
      description: 'Note B Description'
    }
  ]);

  await db('commentmetadata').insert([
    {
      opened: true
    },
    {
      opened: false
    }
  ])

  await db('comment').insert([
    {
      text: 'Note A Comment',
      description: 'Note A Comment Description',
      noteId: 1,
      metadataId: 1
    },
    {
      text: 'Note A Comment 2',
      description: 'Note A Comment Description',
      noteId: 1,
      metadataId: 2
    }
  ]);
}

const getConfig = async () => {
  const config = await loadConfig({
    rootDir: process.cwd(),
    extensions: [
      () => ({ name: 'graphback' })
    ]
  });

  const projectConfig = config.getDefault();
  const graphbackConfig = projectConfig.extension('graphback');

  return { projectConfig, graphbackConfig };
}

test('Find all notes', async () => {
  const { data } = await client.query({ operationName: 'findNotes', query: documents });

  expect(data).toBeDefined();
  expect(data.findNotes).toEqual({
    items: [
      {
        id: '1',
        title: 'Note A',
        description: 'Note A Description',
        comments: [
          {
            id: '1',
            text: 'Note A Comment',
            description: 'Note A Comment Description'
          },
          {
            id: '2',
            text: 'Note A Comment 2',
            description: 'Note A Comment Description'
          }
        ]
      },
      {
        id: '2',
        title: 'Note B',
        description: 'Note B Description',
        comments: []
      }
    ],
    limit: null,
    offset: 0,
    count: 2
  })
})

test('Find all notes except the first', async () => {
  const { data } = await client.query({
    operationName: 'findNotes',
    query: documents,
    variables: { page: { offset: 1 } }
  });

  expect(data).toBeDefined();
  expect(data.findNotes).toEqual({
    items: [
      {
        id: '2',
        title: 'Note B',
        description: 'Note B Description',
        comments: []
      }
    ],
    limit: null,
    offset: 1,
    count: 2
  })
})

test('Find at most one note', async () => {
  const { data } = await client.query({
    operationName: 'findNotes',
    query: documents,
    variables: { page: { limit: 1 } }
  });

  expect(data).toBeDefined();
  expect(data.findNotes).toEqual({
    items: [
      {
        id: '1',
        title: 'Note A',
        description: 'Note A Description',
        comments: [
          {
            id: '1',
            text: 'Note A Comment',
            description: 'Note A Comment Description'
          },
          {
            id: '2',
            text: 'Note A Comment 2',
            description: 'Note A Comment Description'
          }
        ]
      },
    ],
    limit: 1,
    offset: 0,
    count: 2
  })
})

test('Find all comments', async () => {
  const { data } = await client.query({ operationName: "findComments", query: documents });

  expect(data).toBeDefined();
  expect(data.findComments).toEqual({
    items: [
      {
        id: '1',
        text: 'Note A Comment',
        description: 'Note A Comment Description',
        note: {
          id: '1',
          title: 'Note A',
          description: 'Note A Description'
        },
        metadata: {
          id: '1',
          opened: true
        }
      },
      {
        id: '2',
        text: 'Note A Comment 2',
        description: 'Note A Comment Description',
        note: {
          id: '1',
          title: 'Note A',
          description: 'Note A Description'
        },
        metadata: {
          id: '2',
          opened: false
        }
      }
    ],
    limit: null,
    offset: 0,
    count: 2
  })
})

test('Note 1 should be defined', async () => {
  const response = await getNote('1', client);
  expect(response.data).toBeDefined();
  const notes = response.data.getNote;
  expect(notes).toEqual({
    id: '1',
    title: 'Note A',
    description: 'Note A Description',
    comments: [
      {
        id: '1',
        text: 'Note A Comment',
        description: 'Note A Comment Description'
      },
      {
        id: '2',
        text: 'Note A Comment 2',
        description: 'Note A Comment Description'
      }
    ]
  });
})

test('Find at most one comment on Note 1', async () => {
  const response = await client.query({
    operationName: "findComments",
    query: documents,
    variables: { filter: { noteId: { eq: 1 } }, page: { limit: 1 } }
  });

  expect(response.data).toBeDefined()
  const comments = response.data.findComments
  expect(comments.items).toHaveLength(1);
  expect(comments.items).toEqual([
    {
      id: '1',
      text: 'Note A Comment',
      description: 'Note A Comment Description',
      metadata: {
        id: "1",
        opened: true,
      },
      note: {
        description: "Note A Description",
        id: "1",
        title: "Note A",
      },
    }
  ])
})

test('Find comments on Note 1 except first', async () => {
  const response = await client.query({
    operationName: "findComments",
    query: documents,
    variables: { filter: { noteId: { eq: 1 } }, page: { offset: 1 } }
  });

  expect(response.data).toBeDefined()
  const notes = response.data.findComments
  expect(notes.items).toHaveLength(1);
  expect(notes).toEqual({
    items: [
      {
        id: '2',
        text: 'Note A Comment 2',
        description: 'Note A Comment Description',
        metadata: {
          "id": "2",
          "opened": false,
        },
        note: {
          "description": "Note A Description",
          "id": "1",
          "title": "Note A",
        },
      }
    ],
    limit: null,
    offset: 1,
    count: 2
  })
})

test('Should update Note 1 title', async () => {
  const response = await updateNote({ id: '1', title: 'Note 1 New Title' }, client);
  expect(response.data).toBeDefined();
  expect(response.data.updateNote.title).toBe('Note 1 New Title');
});

test('Should create a new Note', async () => {
  const response = await createNote(client, { title: 'New note', description: 'New note description' });
  expect(response.data).toBeDefined();
  expect(response.data.createNote).toEqual({ id: '3', title: 'New note', description: 'New note description' });
})

test('Delete Note 1', async () => {
  const response = await deleteNote(client, '2');
  expect(response.data).toBeDefined();
  expect(response.data.deleteNote).toEqual({ id: '2', description: 'Note B Description', title: 'Note B' });
});

async function updateNote(input: any, client: ApolloServerTestClient) {
  const response = await client.mutate({
    operationName: "updateNote",
    mutation: documents,
    variables: { input }
  });

  return response;
}

async function createNote(client: ApolloServerTestClient, input: any) {
  const response = await client.mutate({
    operationName: "createNote",
    mutation: documents, variables: { input }
  });

  return response;
}

async function deleteNote(client: ApolloServerTestClient, id: string | number) {
  const response = await client.mutate({
    operationName: "deleteNote",
    mutation: documents,
    variables: { input: { id } }
  });

  return response;
}

async function getNote(id: string | number, client: ApolloServerTestClient) {
  const response = await client.query({
    operationName: "getNote",
    query: documents,
    variables: { id }
  });

  return response;
}

async function findNoteComments(noteId: string, client: ApolloServerTestClient) {
  const response = await client.query({
    operationName: "findComments",
    query: documents,
    variables: { filter: { noteId } }
  });

  return response;
}

