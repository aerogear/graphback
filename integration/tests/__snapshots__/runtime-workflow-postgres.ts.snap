// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "tableMap": Map {
    "Note" => Object {
      "annotations": Object {},
      "columnMap": Map {
        "id" => Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": true,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": true,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        "title" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "title",
          "nullable": false,
          "type": "string",
        },
        "description" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
      },
      "columns": Array [
        Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": true,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": true,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "title",
          "nullable": false,
          "type": "string",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
      ],
      "comment": undefined,
      "indexes": Array [],
      "name": "note",
      "uniques": Array [],
    },
    "Comment" => Object {
      "annotations": Object {},
      "columnMap": Map {
        "id" => Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": true,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": true,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        "text" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "text",
          "nullable": true,
          "type": "string",
        },
        "description" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
        "note" => Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "note",
            "type": "Note",
          },
          "isPrimaryKey": false,
          "name": "noteId",
          "nullable": true,
          "type": "integer",
        },
        "metadata" => Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "commentmetadata",
            "type": "CommentMetadata",
          },
          "isPrimaryKey": false,
          "name": "metadataId",
          "nullable": true,
          "type": "integer",
        },
      },
      "columns": Array [
        Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": true,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": true,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "text",
          "nullable": true,
          "type": "string",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
        Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "note",
            "type": "Note",
          },
          "isPrimaryKey": false,
          "name": "noteId",
          "nullable": true,
          "type": "integer",
        },
        Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "commentmetadata",
            "type": "CommentMetadata",
          },
          "isPrimaryKey": false,
          "name": "metadataId",
          "nullable": true,
          "type": "integer",
        },
      ],
      "comment": undefined,
      "indexes": Array [],
      "name": "comment",
      "uniques": Array [],
    },
    "CommentMetadata" => Object {
      "annotations": Object {},
      "columnMap": Map {
        "id" => Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": true,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": true,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        "opened" => Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "opened",
          "nullable": true,
          "type": "boolean",
        },
      },
      "columns": Array [
        Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": true,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": true,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "opened",
          "nullable": true,
          "type": "boolean",
        },
      ],
      "comment": undefined,
      "indexes": Array [],
      "name": "commentmetadata",
      "uniques": Array [],
    },
  },
  "tables": Array [
    Object {
      "annotations": Object {},
      "columnMap": Map {
        "id" => Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": true,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": true,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        "title" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "title",
          "nullable": false,
          "type": "string",
        },
        "description" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
      },
      "columns": Array [
        Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": true,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": true,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "title",
          "nullable": false,
          "type": "string",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
      ],
      "comment": undefined,
      "indexes": Array [],
      "name": "note",
      "uniques": Array [],
    },
    Object {
      "annotations": Object {},
      "columnMap": Map {
        "id" => Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": true,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": true,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        "text" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "text",
          "nullable": true,
          "type": "string",
        },
        "description" => Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
        "note" => Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "note",
            "type": "Note",
          },
          "isPrimaryKey": false,
          "name": "noteId",
          "nullable": true,
          "type": "integer",
        },
        "metadata" => Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "commentmetadata",
            "type": "CommentMetadata",
          },
          "isPrimaryKey": false,
          "name": "metadataId",
          "nullable": true,
          "type": "integer",
        },
      },
      "columns": Array [
        Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": true,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": true,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "text",
          "nullable": true,
          "type": "string",
        },
        Object {
          "annotations": Object {},
          "args": Array [
            255,
          ],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "description",
          "nullable": true,
          "type": "string",
        },
        Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "note",
            "type": "Note",
          },
          "isPrimaryKey": false,
          "name": "noteId",
          "nullable": true,
          "type": "integer",
        },
        Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": Object {
            "columnName": "id",
            "field": "id",
            "tableName": "commentmetadata",
            "type": "CommentMetadata",
          },
          "isPrimaryKey": false,
          "name": "metadataId",
          "nullable": true,
          "type": "integer",
        },
      ],
      "comment": undefined,
      "indexes": Array [],
      "name": "comment",
      "uniques": Array [],
    },
    Object {
      "annotations": Object {},
      "columnMap": Map {
        "id" => Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": true,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": true,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        "opened" => Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "opened",
          "nullable": true,
          "type": "boolean",
        },
      },
      "columns": Array [
        Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": true,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": true,
          "name": "id",
          "nullable": false,
          "type": "increments",
        },
        Object {
          "annotations": Object {},
          "args": Array [],
          "autoIncrementable": false,
          "comment": undefined,
          "defaultValue": undefined,
          "foreign": undefined,
          "isPrimaryKey": false,
          "name": "opened",
          "nullable": true,
          "type": "boolean",
        },
      ],
      "comment": undefined,
      "indexes": Array [],
      "name": "commentmetadata",
      "uniques": Array [],
    },
  ],
}
`;
